### PR TITLE
[RLlib] Trajectory view API - 02 actual API scaffold

### PR DESCRIPTION
**Pull request by sven1977**
This PR introduces the actual API (class) that will be used by Samplers to store data and to retrieve it again for:
- action computations 
- postprocess batches
- train batches

No functionality is actually being switched on in the PR.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/evaluation/sample_collector.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                agents.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def add_to_forward_pass(self,
+                            episode_id: EpisodeID,
+                            agent_id: AgentID,
+                            env_id: EnvID) -> None:
+        """Registers given agent/episode combination for the next forward pass.
```
Comment (ericl) [inline rllib/evaluation/sample_collector.py:82]: It's a bit confusing how this is different from add init obs. Maybe add some examples on how to use this collector interface?

File: rllib/evaluation/sample_collector.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                registering for the upcoming forward pass.
+            agent_id (AgentID): Unique id for the agent we are registering
+                for the upcoming forward pass.
+            env_id (EnvID): Unique id for the (sub)-environment.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def reset_forward_pass(self) -> None:
+        """Resets all forward pass information (new items can be registered).
```
Comment (ericl) [inline rllib/evaluation/sample_collector.py:95]: When would you use this?

File: rllib/evaluation/sample_collector.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    Note: This is an experimental class only used when
+    `config._use_trajectory_view_api` = True.
+    Once `_use_trajectory_view_api` becomes the default in configs:
+    This class will deprecate the `SampleBatchBuilder` and
+    `MultiAgentBatchBuilder` classes.
+
+    This API is controlled by RolloutWorker objects to store all data
+    generated by Environments and Policies/Models during rollout and
+    postprocessing. Also provides a method to generate input_dicts for
+    action calculation and train batches from the collected data.
```
Comment (ericl) [inline rllib/evaluation/sample_collector.py:25]: The comment could be improved to better motivate the reason for this method (to provide performance benefits).

File: rllib/models/modelv2.py [hunk 4]
```diff
@@ -246,35 +246,24 @@ def from_batch(self, train_batch: SampleBatch,
...
         """
         # Default implementation for simple RL model:
         # Single requirement: Pass current obs as input.
         return {
-            SampleBatch.CUR_OBS: ViewRequirement(timesteps=0),
-            SampleBatch.PREV_ACTIONS: ViewRequirement(
-                SampleBatch.ACTIONS, timesteps=-1),
-            SampleBatch.PREV_REWARDS: ViewRequirement(
-                SampleBatch.REWARDS, timesteps=-1),
+            SampleBatch.OBS: ViewRequirement(),
```
Comment (ericl) [inline rllib/models/modelv2.py:266]: Does it make sense to pass timesteps as well? `timesteps=1`?

File: rllib/policy/torch_policy.py [hunk 5]
```diff
@@ -103,6 +103,11 @@ def __init__(
         else:
             self.device = torch.device("cpu")
         self.model = model.to(self.device)
+        # Combine view_requirements for Model and Policy.
+        self.view_requirements = {
+            **self.model.get_view_requirements(),
+            **self.get_view_requirements(),
+        }
```
Comment (ericl) [inline rllib/policy/torch_policy.py:110]: Nice!

File: rllib/policy/torch_policy.py [hunk 6]
```diff
@@ -120,6 +125,17 @@ def __init__(
             get_batch_divisibility_req(self) if get_batch_divisibility_req \
             else 1
 
+    @override(Policy)
+    def get_view_requirements(self):
```
Comment (ericl) [inline rllib/policy/torch_policy.py:129]: For clarity, should we rename these to `get_training_view_requirements` and `get_inference_view_requirements`? Or perhaps policy view / model view? I suppose inference view is always a subset of training view?

File: rllib/policy/view_requirement.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    name (C) under which the view is available in the `input_dict` and
+    ViewRequirement specifies the actual underlying column names (in the
+    original data buffer), timesteps, and other options to build the view
+    for N.
+
+    Examples:
+        >>> # The default ViewRequirement for a Model is:
+        >>> req = [ModelV2].get_view_requirements(is_training=False)
+        >>> print(req)
+        {"obs": ViewRequirement(timesteps=0)}
```
Comment (ericl) [inline rllib/policy/view_requirement.py:25]: Do we still have timesteps? Also, it makes sense to have timesteps=1 mean one timestep right?

File: rllib/policy/view_requirement.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    """
+
+    def __init__(self,
+                 data_col: Optional[str] = None,
+                 space: Space = None,
```
Comment (ericl) [inline rllib/policy/view_requirement.py:30]: gym.Space for clarity

File: rllib/policy/view_requirement.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+                 shift: int = 0,
+                 sampling: bool = True,
```
Comment (ericl) [inline rllib/policy/view_requirement.py:32]: Let's try to remove these "when is the view requirement used" flags from the view req. Instead, we should keep a data structure that has separate lists of which view requirements are used in which scenarios.

File: rllib/policy/view_requirement.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+                If None, use the dict key under which this ViewRequirement
+                resides.
+            space (Space): The gym Space used in case we need to pad data in
+                unaccessible areas of the trajectory (t<0 or t>H).
+                Default: Simple box space, e.g. rewards.
+            shift (Union[List[int], int]): List of relative (or absolute
+                timesteps) to be present in the input_dict.
+            repeat_mode (str): The repeat-mode (one of "all" or "only_first").
+                E.g. for training, we only want the first internal state
+                timestep (the NN will calculate all others again anyways).
```
Comment (ericl) [inline rllib/policy/view_requirement.py:48]: Docs seem out of sync here.

**Review by ericl — CHANGES_REQUESTED** [review]

File: rllib/evaluation/sample_batch_builder.py [hunk 11]
```diff
@@ -29,6 +29,9 @@ def to_float_array(v: List[Any]) -> np.ndarray:
 class SampleBatchBuilder:
     """Util to build a SampleBatch incrementally.
 
+    Note: This will be deprecated by _FastMultiAgentSampleBatchBuilder in the
+    future.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/sample_batch_builder.py:33]: done

File: rllib/policy/view_requirement.py (hunk 7 above)
Comment (sven1977) [reply rllib/policy/view_requirement.py:25]: example was fixed

File: rllib/policy/view_requirement.py (hunk 8 above)
Comment (sven1977) [reply rllib/policy/view_requirement.py:30]: done

File: rllib/policy/view_requirement.py (hunk 9 above)
Comment (sven1977) [reply rllib/policy/view_requirement.py:32]: ok, see my answer above, too on the "when is the view requirement used" flags.

File: rllib/evaluation/sample_collector.py (hunk 3 above)
Comment (sven1977) [reply rllib/evaluation/sample_collector.py:25]: done

File: rllib/evaluation/sample_collector.py (hunk 1 above)
Comment (sven1977) [reply rllib/evaluation/sample_collector.py:82]: done

File: rllib/evaluation/sample_collector.py (hunk 2 above)
Comment (sven1977) [reply rllib/evaluation/sample_collector.py:95]: I'll add more comments/examples. It's basically to wipe out stored index information for the next forward pass (which agent(s)/timestep(s) to select from the buffer to form the next inputs?).

File: rllib/agents/ppo/ppo_torch_policy.py [hunk 12]
```diff
@@ -216,6 +217,16 @@ def setup_mixins(policy, obs_space, action_space, config):
     LearningRateSchedule.__init__(policy, config["lr"], config["lr_schedule"])
 
 
+def get_view_requirements_fn(policy):
+    return {
+        # Next obs are needed for PPO postprocessing.
+        SampleBatch.NEXT_OBS: ViewRequirement(
+            SampleBatch.OBS, shift=1, sampling=False, training=False),
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/ppo_torch_policy.py:224]: ok, let's just separate into model.get_view_requirements() (all forward passes) and policy.get_view_requirements() (includes model requirements plus needed postprocessing/loss views). See also discussion below.

File: rllib/policy/view_requirement.py (hunk 10 above)
Comment (sven1977) [reply rllib/policy/view_requirement.py:48]: done

File: rllib/policy/torch_policy.py (hunk 6 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py:129]: Yes, it's actually that simple:
Model.get_view_requirements() -> always for forward pass.
Policy.get_view_requirements() -> can add additional fields required for postprocessing and loss.
I'll fix these inconsistencies.

File: rllib/models/modelv2.py (hunk 4 above)
Comment (sven1977) [reply rllib/models/modelv2.py:266]: actually, timesteps is obsolete. Will fix this.

File: rllib/agents/ppo/ppo_torch_policy.py [hunk 13]
```diff
@@ -229,4 +239,6 @@ def setup_mixins(policy, obs_space, action_space, config):
     mixins=[
         LearningRateSchedule, EntropyCoeffSchedule, KLCoeffMixin,
         ValueNetworkMixin
-    ])
+    ],
+    get_view_requirements=get_view_requirements_fn,
```
Comment (ericl) [inline rllib/agents/ppo/ppo_torch_policy.py:243]:
```suggestion
    policy_view_requirements=get_view_requirements_fn,
```

File: rllib/evaluation/sample_batch_builder.py [hunk 14]
```diff
@@ -29,6 +29,8 @@ def to_float_array(v: List[Any]) -> np.ndarray:
 class SampleBatchBuilder:
     """Util to build a SampleBatch incrementally.
 
+    TODO(sven): remove this once we switch to trajectory view API.
```
Comment (ericl) [inline rllib/evaluation/sample_batch_builder.py:32]: Put TODOs as normal python comments, not in the pydoc.

File: rllib/evaluation/sample_batch_builder.py [hunk 15]
```diff
@@ -76,6 +78,8 @@ def build_and_reset(self) -> SampleBatch:
 class MultiAgentSampleBatchBuilder:
     """Util to build SampleBatches for each policy in a multi-agent env.
 
+    TODO(sven): remove this once we switch to trajectory view API.
```
Comment (ericl) [inline rllib/evaluation/sample_batch_builder.py:81]: Make it a normal python comment.

**Review by ericl — COMMENTED** [review]
One thing that could use more clarification is policy vs model view requirements. How about we call these `inference_view_requirements` and `training_view_requirements`? I think that removes all possible ambiguity. This means renaming all generic `get_view_requirement` methods to one of these two names.

File: rllib/evaluation/sample_batch_builder.py (hunk 14 above)
Comment (sven1977) [reply rllib/evaluation/sample_batch_builder.py:32]: done

File: rllib/evaluation/sample_batch_builder.py (hunk 15 above)
Comment (sven1977) [reply rllib/evaluation/sample_batch_builder.py:81]: done

File: rllib/agents/ppo/ppo_torch_policy.py (hunk 13 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo_torch_policy.py:243]: done

File: rllib/evaluation/sample_collector.py (hunk 2 above)
Comment (sven1977) [reply rllib/evaluation/sample_collector.py]: I actually took it out. I don't think it's needed.

File: rllib/policy/torch_policy.py [hunk 16]
```diff
@@ -116,9 +121,20 @@ def __init__(
...
-            get_batch_divisibility_req(self) if get_batch_divisibility_req \
-            else 1
+        self.batch_divisibility_req = get_batch_divisibility_req(self) if \
+            callable(get_batch_divisibility_req) else \
+            (get_batch_divisibility_req or 1)
+
+    @override(Policy)
+    def training_view_requirements(self):
+        if hasattr(self, "view_requirements"):
+            return self.view_requirements
```
Comment (ericl) [inline rllib/policy/torch_policy.py:131]: Isn't this always the case since we set this in the constructor?

**Review by ericl — APPROVED** [review]
LGTM just one question

**Pull request merged**